### ISSUE TITLE
"Failed to launch: fork/exec /bin/sh: operation not permitted" when starting a session on a mac node

### ISSUE THREAD

**Issue by alex-kovoy** [gravitational/teleport#5035]
We see an error when trying to start a session on a mac node (joined via server-auto-install script) :
```
Failed to launch: fork/exec /bin/sh: operation not permitted.
```

This error is shown during the very first session start attempt. It works fine on consequent attempts.

**Comment by webvictim** [comment]
I've seen this issue too, it also happens with `/bin/zsh`.

**Comment by russjones** [comment]
@alex-kovoy What version of macOS does this occur on?

**Comment by webvictim** [comment]
Here's some logs from when it's happened to me:
```
seamini:~ root# tail /var/log/teleport-stderr.log
ERRO [NODE]      Unable to run shell command (/bin/sh): fork/exec /bin/sh: operation not permitted id:1 local:10.0.0.141:49177 login:root remote:1.2.3.4:51653 teleportUser:redacted@example.com srv/sess.go:607
ERRO [NODE]      fork/exec /bin/sh: operation not permitted regular/sshserver.go:1389
ERRO [NODE]      Unable to run shell command (/bin/sh): fork/exec /bin/sh: operation not permitted id:1 local:10.0.0.141:62732 login:root remote:1.2.3.4:52861 teleportUser:redacted@example.com srv/sess.go:607
ERRO [NODE]      fork/exec /bin/sh: operation not permitted regular/sshserver.go:1389

seamini:~ root# sw_vers
ProductName:    Mac OS X
ProductVersion: 10.15.7
BuildVersion:   19H15
```

**Comment by klizhentas** [comment]
@russjones I reproduce it 100% with BigSur and cloud, can share any logs or access if you want.

**Comment by webvictim** [comment]
https://walac.github.io/golang-patch/ seems similarish.

**Comment by webvictim** [comment]
I've noticed that whenever this happens, it's always on the first connection to the node after Teleport has been started. On the second connection, everything works fine until Teleport restarts again.

**Comment by webvictim** [comment]
Just noticed this again setting up a new build box. The shell fails to spawn the first time after restarting the Teleport process, but then works fine the second time.

Related: https://github.com/gravitational/teleport/issues/3167

**Comment by webvictim** [comment]
This is still a problem in Teleport 8.0.6.

Here's a full set of logs from the node when trying to log in:

```
2021-12-16T12:11:44-04:00 DEBU [NODE:PROX] Transport request: teleport-transport. leaseID:1 target:teleport.example.com:3024 reversetunnel/agent.go:483
2021-12-16T12:11:44-04:00 DEBU [NODE:PROX] Received out-of-band proxy transport request for @local-node [f4f1bb7c-e873-43e5-b0a0-6c1e558675b1.teleport.example.com]. leaseID:1 target:teleport.example.com:3024 reversetunnel/transport.go:195
2021-12-16T12:11:44-04:00 DEBU [NODE:PROX] Handing off connection to a local "node" service. leaseID:1 target:teleport.example.com:3024 reversetunnel/transport.go:270
2021-12-16T12:11:44-04:00 DEBU [NODE]      conn(192.168.64.200:52048->192.168.64.17:57419, user=gus) auth attempt fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:paYrabEKnN0UxysM3dq1A6RjR7OckqARoCYoQsy0ftQ local:192.168.64.17:57419 remote:192.168.64.200:52048 user:gus srv/authhandlers.go:211
2021-12-16T12:11:44-04:00 DEBU [NODE]      conn(192.168.64.200:52048->192.168.64.17:57419, user=gus) auth attempt with key ssh-rsa-cert-v01@openssh.com SHA256:paYrabEKnN0UxysM3dq1A6RjR7OckqARoCYoQsy0ftQ, &ssh.Certificate{Nonce:[]uint8{0xaa, 0xcf, 0x7b, 0xc, 0x2, 0xdb, 0x8b, 0xfc, 0xc0, 0x2d, 0x90, 0xd9, 0x8c, 0x83, 0xad, 0xcb, 0x62, 0xe4, 0x2b, 0x5e, 0xdf, 0x6b, 0x50, 0xdf, 0x80, 0xcf, 0xa, 0xd9, 0xa, 0xe6, 0xdc, 0x87}, Key:(*ssh.rsaPublicKey)(0x14000b12b80), Serial:0x0, CertType:0x1, KeyId:"webvictim", ValidPrincipals:[]string{"gus", "webvictim", "ec2-user", "bots"}, ValidAfter:0x61bb63bb, ValidBefore:0x61bc0b06, Permissions:ssh.Permissions{CriticalOptions:map[string]string{}, Extensions:map[string]string{"permit-X11-forwarding":"", "permit-agent-forwarding":"", "permit-port-forwarding":"", "permit-pty":"", "teleport-roles":"{\"version\":\"v1\",\"roles\":[\"admin\"]}", "teleport-route-to-cluster":"teleport.example.com", "teleport-traits":"{\"kubernetes_groups\":null,\"kubernetes_users\":null,\"logins\":[\"webvictim\"]}"}}, Reserved:[]uint8{}, SignatureKey:(*ssh.rsaPublicKey)(0x14000b12bc0), Signature:(*ssh.Signature)(0x140010facc0)} fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:paYrabEKnN0UxysM3dq1A6RjR7OckqARoCYoQsy0ftQ local:192.168.64.17:57419 remote:192.168.64.200:52048 user:gus srv/authhandlers.go:214
2021-12-16T12:11:44-04:00 DEBU [NODE]      Successfully authenticated fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:paYrabEKnN0UxysM3dq1A6RjR7OckqARoCYoQsy0ftQ local:192.168.64.17:57419 remote:192.168.64.200:52048 user:gus srv/authhandlers.go:274
2021-12-16T12:11:44-04:00 DEBU [NODE]      Checking permissions for (webvictim,gus) to login to node with RBAC checks. srv/authhandlers.go:403
2021-12-16T12:11:44-04:00 DEBU [RBAC]      Access to node "f4f1bb7c-e873-43e5-b0a0-6c1e558675b1" granted, allow rule in role "admin" matched. services/role.go:1650
2021-12-16T12:11:44-04:00 DEBU [SSH:NODE]  Incoming connection 192.168.64.200:52048 -> 192.168.64.17:57419 version: SSH-2.0-Go, certtype: "user" sshutils/server.go:466
2021-12-16T12:11:44-04:00 DEBU [SSH:NODE]  Received out-of-band request: &{Type:x-teleport-version WantReply:true Payload:[] ch:<nil> mux:0x140008f4000}. sshutils/server.go:521
2021-12-16T12:11:44-04:00 DEBU [KEEPALIVE] Starting keep-alive loop with with interval 5m0s and max count 3. srv/keepalive.go:65
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 DEBU             Will create new session for SSH connection 192.168.64.200:52048. srv/ctx.go:482
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request auth-agent-req@openssh.com, want reply true. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 DEBU [NODE]      Starting agent server for Teleport user webvictim and socket /tmp/teleport-2953568290/teleport-61569.socket. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:881
2021-12-16T12:11:44-04:00 DEBU             Will create new session for SSH connection 192.168.64.200:52048. srv/ctx.go:482
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request pty-req, want reply true. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 DEBU [NODE]      Requested terminal "xterm" of size {80 25} id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim srv/termhandlers.go:73
2021-12-16T12:11:44-04:00 DEBU             Set permissions on /dev/ttys002 to 501:4 with mode -rw-------. srv/term.go:407
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Unable to update window size, no session found in context. srv/sess.go:392
2021-12-16T12:11:44-04:00 DEBU             Will create new session for SSH connection 192.168.64.200:52048. srv/ctx.go:482
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request shell, want reply true. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 INFO [NODE]      Creating (interactive) session c7d50885-1a7d-4042-bcf9-c0943d208553. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim srv/sess.go:213
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Using async streamer for session c7d50885-1a7d-4042-bcf9-c0943d208553. srv/sess.go:1066
2021-12-16T12:11:44-04:00 INFO [SESSION:N] New party ServerContext(192.168.64.200:52048->192.168.64.17:57419, user=gus, id=1) party(id=30fee7bf-986c-4217-aeda-46e069e5fa97) joined session: c7d50885-1a7d-4042-bcf9-c0943d208553 srv/sess.go:1242
2021-12-16T12:11:44-04:00 DEBU             Will join session c7d50885-1a7d-4042-bcf9-c0943d208553 for SSH connection 192.168.64.200:52048. srv/ctx.go:484
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Starting poll and sync of terminal size to all parties. srv/sess.go:1184
2021-12-16T12:11:44-04:00 INFO [AUDIT]     session.start addr.remote:192.168.64.200:52048 cluster_name:teleport.example.com code:T2000I ei:0 event:session.start login:gus namespace:default server_addr:192.168.64.17:57419 server_hostname:athena server_id:f4f1bb7c-e873-43e5-b0a0-6c1e558675b1 server_labels:map[arch:arm64 external_ipv4:142.167.142.235 external_ipv6:No IPv6 hostname:athena.home.lan teleport_version:Teleport v8.0.6 git:v8.0.6-0-g807c6d662 go1.17.5 type:node] session_recording:node sid:c7d50885-1a7d-4042-bcf9-c0943d208553 size:80:25 time:2021-12-16T16:11:44.526Z uid:6cc7e5d4-9323-4fa1-bbeb-fa479917f304 user:webvictim events/emitter.go:324
2021-12-16T12:11:44-04:00 DEBU [NODE]      Handling request window-change, want reply false. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1402
2021-12-16T12:11:44-04:00 DEBU             Will join session c7d50885-1a7d-4042-bcf9-c0943d208553 for SSH connection 192.168.64.200:52048. srv/ctx.go:484
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Copying from PTY to writer completed with error <nil>. srv/sess.go:800
2021-12-16T12:11:44-04:00 DEBU [NODE]      Exec request ("/usr/local/bin/teleport") complete: 255 id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1382
2021-12-16T12:11:44-04:00 INFO [SESSION:N] Closing party 30fee7bf-986c-4217-aeda-46e069e5fa97 srv/sess.go:1421
2021-12-16T12:11:44-04:00 DEBU [NODE]      Releasing associated resources - context has been closed. id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim srv/monitor.go:251
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Sent session.join to 192.168.64.200:52048. srv/sess.go:292
2021-12-16T12:11:44-04:00 INFO [NODE]      Removing party ServerContext(192.168.64.200:52048->192.168.64.17:57419, user=gus, id=1) party(id=30fee7bf-986c-4217-aeda-46e069e5fa97) from session c7d50885-1a7d-4042-bcf9-c0943d208553 id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim srv/sess.go:1106
2021-12-16T12:11:44-04:00 INFO [AUDIT]     session.leave cluster_name:teleport.example.com code:T2003I ei:3 event:session.leave namespace:default server_addr:192.168.64.17:57419 server_hostname:athena server_id:f4f1bb7c-e873-43e5-b0a0-6c1e558675b1 server_labels:map[arch:arm64 external_ipv4:142.167.142.235 external_ipv6:No IPv6 hostname:athena.home.lan teleport_version:Teleport v8.0.6 git:v8.0.6-0-g807c6d662 go1.17.5 type:node] sid:30fee7bf-986c-4217-aeda-46e069e5fa97 time:2021-12-16T16:11:44.583Z uid:3826dcb4-1169-47eb-8b69-612019538c7c user:webvictim events/emitter.go:324
2021-12-16T12:11:44-04:00 INFO [SESSION:N] Session c7d50885-1a7d-4042-bcf9-c0943d208553 will be garbage collected. srv/sess.go:325
2021-12-16T12:11:44-04:00 INFO [SESSION:N] Closing session c7d50885-1a7d-4042-bcf9-c0943d208553. srv/sess.go:640
2021-12-16T12:11:44-04:00 DEBU [SESSION:N] Stopping poll and sync of terminal size to all parties. srv/sess.go:1205
2021-12-16T12:11:44-04:00 INFO [SESSION:N] Party member 30fee7bf-986c-4217-aeda-46e069e5fa97 left session c7d50885-1a7d-4042-bcf9-c0943d208553. srv/sess.go:1251
2021-12-16T12:11:44-04:00 DEBU [SSH:NODE]  Closed connection 192.168.64.200:52048. sshutils/server.go:471
2021-12-16T12:11:44-04:00 DEBU             AgentServer(/tmp/teleport-2953568290/teleport-61569.socket) is closing teleagent/agent.go:156
2021-12-16T12:11:44-04:00 DEBU [TERM:LOCA] Closed PTY srv/term.go:300
2021-12-16T12:11:44-04:00 INFO [AUDIT]     session.data addr.remote:192.168.64.200:52048 code:T2006I ei:2.147483646e+09 event:session.data login:gus namespace:default rx:4702 server_id:f4f1bb7c-e873-43e5-b0a0-6c1e558675b1 sid:c7d50885-1a7d-4042-bcf9-c0943d208553 time:2021-12-16T16:11:44.595Z tx:4720 uid:e81f46ff-8919-4e21-9e25-6d0965d695cd user:webvictim events/emitter.go:324
2021-12-16T12:11:44-04:00 INFO [AUDIT]     session.end cluster_name:teleport.example.com code:T2004I ei:4 enhanced_recording:false event:session.end interactive:true namespace:default participants:[webvictim] server_addr:192.168.64.17:57419 server_hostname:athena server_id:f4f1bb7c-e873-43e5-b0a0-6c1e558675b1 server_labels:map[arch:arm64 external_ipv4:142.167.142.235 external_ipv6:No IPv6 hostname:athena.home.lan teleport_version:Teleport v8.0.6 git:v8.0.6-0-g807c6d662 go1.17.5 type:node] session_recording:node session_start:2021-12-16T16:11:44.50887Z session_stop:2021-12-16T16:11:44.583539Z sid:c7d50885-1a7d-4042-bcf9-c0943d208553 time:2021-12-16T16:11:44.584Z uid:c439c89e-5673-4231-a28f-c4dd86d51e6d user:webvictim events/emitter.go:324
```

This seems relevant:

```
2021-12-16T12:11:44-04:00 DEBU [NODE]      Exec request ("/usr/local/bin/teleport") complete: 255 id:1 local:192.168.64.17:57419 login:gus remote:192.168.64.200:52048 teleportUser:webvictim regular/sshserver.go:1382
```

**Comment by ibeckermayer** [comment]
I just encountered this myself:

`macOS Ventura 13.0`
`Teleport v12.0.0-dev git:v9.3.18-40-geff460b19 go1.19.1`

The issue wasn't solve on consequent attempts, but could be solved by running teleport as `sudo`

**Comment by espadolini** [comment]
The issue that I was able to find shows up as
```
Failed to launch: fork/exec /bin/zsh: invalid argument.
ERROR: Process exited with status 255
```
when doing `tsh ssh` as any host user other than the user running the `teleport` binary.

I traced it to `setgroups` in darwin outright failing when given more than `NGROUPS_MAX` (16) groups; `setgroups(2)` states: `Use of setgroups() is highly discouraged.`

From some quick testing, simply not setting `syscall.Credentials.Groups` in `lib/srv.getCmdCredential` when the OS is darwin fixes the issue, and the list of groups in the opened shells (as displayed by `id`/`id -G`) seems to be correct for both my local user and for `root`. I will investigate further, but it seems to be the simplest way to fix the bug, especially considering that we still get the correct behavior by doing less work.
